pkg/apis/resource/validation/validation.go

```diff
@@ -433,179 +512,244 @@ func validateNodeName(name string, fldPath *field.Path) field.ErrorList {
+	if combinedLen, max := len(device.Attributes)+len(device.Capacity), resource.ResourceSliceMaxAttributesAndCapacitiesPerDevice; combinedLen > max {
+		allErrs = append(allErrs, field.Invalid(fldPath, combinedLen, fmt.Sprintf("the total number of attributes and capacities must not exceed %d", max)))
 	}
 	return allErrs
 }
 
-func validateResourceRequestModel(model *resource.ResourceRequestModel, fldPath *field.Path, requestStored bool) field.ErrorList {
+var (
+	numericIdentifier = `(0|[1-9]\d*)`
+
+	preReleaseIdentifier = `(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)`
+
+	buildIdentifier = `[0-9a-zA-Z-]+`
+
+	semverRe = regexp.MustCompile(`^` +
+
+		// dot-separated version segments (e.g. 1.2.3)
+		numericIdentifier + `\.` + numericIdentifier + `\.` + numericIdentifier +
+
+		// optional dot-separated prerelease segments (e.g. -alpha.PRERELEASE.1)
+		`(-` + preReleaseIdentifier + `(\.` + preReleaseIdentifier + `)*)?` +
+
+		// optional dot-separated build identifier segments (e.g. +build.id.20240305)
+		`(\+` + buildIdentifier + `(\.` + buildIdentifier + `)*)?` +
+
+		`$`)
+)
+
+func validateDeviceAttribute(attribute resource.DeviceAttribute, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
-	entries := sets.New[string]()
-	if model.NamedResources != nil {
-		entries.Insert("namedResources")
-		allErrs = append(allErrs, namedresourcesvalidation.ValidateRequest(namedresourcesvalidation.Options{StoredExpressions: requestStored}, model.NamedResources, fldPath.Child("namedResources"))...)
+	numFields := 0
+	if attribute.BoolValue != nil {
+		numFields++
+	}
+	if attribute.IntValue != nil {
+		numFields++
+	}
+	if attribute.StringValue != nil {
+		if len(*attribute.StringValue) > resource.DeviceAttributeMaxValueLength {
+			allErrs = append(allErrs, field.TooLongMaxLength(fldPath.Child("string"), *attribute.StringValue, resource.DeviceAttributeMaxValueLength))
+		}
+		numFields++
 	}
-	switch len(entries) {
+	if attribute.VersionValue != nil {
+		numFields++
+		if !semverRe.MatchString(*attribute.VersionValue) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("version"), *attribute.VersionValue, "must be a string compatible with semver.org spec 2.0.0"))
+		}
+		if len(*attribute.VersionValue) > resource.DeviceAttributeMaxValueLength {
+			allErrs = append(allErrs, field.TooLongMaxLength(fldPath.Child("version"), *attribute.VersionValue, resource.DeviceAttributeMaxValueLength))
+		}
+	}
+
+	switch numFields {
 	case 0:
-		allErrs = append(allErrs, field.Required(fldPath, "exactly one structured model field must be set"))
+		allErrs = append(allErrs, field.Required(fldPath, "exactly one value must be set"))
 	case 1:
 		// Okay.
 	default:
-		allErrs = append(allErrs, field.Invalid(fldPath, sets.List(entries), "exactly one field must be set, not several"))
+		allErrs = append(allErrs, field.Invalid(fldPath, attribute, "exactly one field must be set, not several"))
```

Comment:
Both cases can use the same string- "exactly one value must be specified" (not "set")

Reply:
Ack (to both).